Improve exception when HTTP.sys can't register a prefix

0xced opened the pull request:
Propose an actionable solution and include a link to the documentation.

@guardrex  Can you create an aka.ms link to https://docs.microsoft.com/aspnet/core/fundamentals/servers/httpsys#configure-windows-server that I could use in the exception message?

guardrex commented:
I can't ... I don't have access ... but @Rick-Anderson might help or advise.

halter73 commented on src/Servers/HttpSys/src/NativeInterop/UrlGroup.cs:
```diff
@@ -79,12 +79,13 @@ internal void RegisterPrefix(string uriPrefix, int contextId)
+                if (statusCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_ACCESS_DENIED)
                 {
-                    throw new HttpSysException((int)statusCode);
+                    throw new HttpSysException((int)statusCode, Resources.FormatException_AccessDenied(uriPrefix, Environment.UserName));
```
Can we add a test for this?

Tratcher replied:
Is Environment.UserName in the right format to insert directly into the command?

0xced replied:
@halter73 I added a test in 8784857099fa085913ef4c4d03dd3e7743d1db30.
@Tratcher I included both the domain + user in `DOMAIN\User` format, as documented on [Configuring namespace reservations](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/configuring-http-and-https#configuring-namespace-reservations), also in 8784857099fa085913ef4c4d03dd3e7743d1db30.

halter73 commented on src/Servers/HttpSys/src/Resources.resx:
```diff
@@ -117,6 +117,11 @@
+  <data name="Exception_AccessDenied" xml:space="preserve">
+    <value>The prefix '{0}' is not registered. Please run the following command as Administrator to register this prefix:
+netsh http add urlacl url={0} user={1}
+See "Preregister URL prefixes on the server" on https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/httpsys#configure-windows-server for more information.</value>
```
@shirhatti Can you create an aka.ms link for this?

shirhatti commented on src/Servers/HttpSys/src/Resources.resx:
```diff
@@ -117,6 +117,11 @@
+  <data name="Exception_AccessDenied" xml:space="preserve">
+    <value>The prefix '{0}' is not registered. Please run the following command as Administrator to register this prefix:
+netsh http add urlacl url={0} user={1}
+See "Preregister URL prefixes on the server" on https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/httpsys#configure-windows-server for more information.</value>
```
```suggestion
See "Preregister URL prefixes on the server" on https://go.microsoft.com/fwlink/?linkid=2127065 for more information.</value>
```

0xced replied:
Done, thanks for creating the link @shirhatti.

0xced committed:
Improve exception when HTTP.sys can't register a prefix

Propose an actionable solution and include a link to the documentation.

0xced commented:
I addressed the concerns raised by @shirhatti, @halter73 and @Tratcher, rebased onto master and force-pushed 52c0effaa8130897688f740069f69edc2f4a6bfb.

Tratcher commented:
The new test failed on helix.
https://dev.azure.com/dnceng/public/_build/results?buildId=614229&view=ms.vss-test-web.build-test-results-tab&runId=19274364&resultId=107864&paneView=debug
```
Assert.Throws() Failure

Expected: typeof(Microsoft.AspNetCore.Server.HttpSys.HttpSysException)

Actual: (No exception was thrown)
```

You wouldn't get AccessDenied if the test were run as admin, right? I wonder if Helix does that.

shirhatti commented:
You could change the hostname in the test to localhost though? You wouldn't need elevation for that

Tratcher commented:
The point is that the test wants AccessDenied and isn't getting it.

HaoK commented:
So this is failing because of a sudo issue?

HaoK commented:
Yeah so our helix jobs do run as admin, so if the tests were expecting admin access failures, you might need to skip this test on helix

0xced committed:
Add SkipOnHelix attribute

HaoK commented on src/Servers/HttpSys/test/FunctionalTests/Listener/ServerTests.cs:
```diff
@@ -123,6 +125,20 @@ public async Task Server_SetRejectionVerbosityLevel_Success()
         }
 
+        [ConditionalFact]
+        [SkipOnHelix("https://github.com/dotnet/aspnetcore/pull/20718#issuecomment-618758634")]
```
You can target specific queues, if this only is an issue on Windows queues, it might be best to use the OS skip on windows

Tratcher replied:
All HttpSys tests only run on Windows.

Tratcher commented:
Unrelated broken test: https://github.com/dotnet/aspnetcore/pull/21191
